keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -118,78 +142,79 @@ Currently, kubeadm vendors in pieces of Go code from the components to be able t
+The loaded configs, that were supplied with `--config` would be taken into account when printing the additional component config information.
 
-If a user has provided Kustomize patches for a config version, that is different from the current one, kubeadm is going to bail out with an error, prompting the user to either delete the old patches, or to upgrade them to the new version. This is best implemented using a kubeadm pre-flight check.
+### "kubeadm upgrade apply" changes
 
-UX wise, this feature is going to be reusing the existing kubeadm options that were introduced with the implementation of the kubeadm advanced configurations KEP. This allows for kustomizations to be applied to component configs on a per-node basis during init, join and upgrade just before the config is supposed to be stored on the local node file.
+As supported component configs are expected to increase in number in the future and some of them might require manual user migration prior to upgrade, a handy and secure method for supplying new configs is necessary.
 
-### Stop using component config internal types
+For that matter `kubeadm upgrade apply` also gains the same extended semantics of the `--config` flag.  
+In addition to that, the newly used configs, from the file supplied with `--config`, would be stored in the cluster as user specified configs during the upgrade process.
 
-As kubeadm will no longer default, validate or migrate component configs inside its code base, the need to use component config internal types is no longer needed. In that case the usage of internal types can be removed. This would also allow us to reduce the number of imported packages and to drop one of the last dependencies on the kubernetes/kubernetes code base.  
-The switch is going to be performed by replacing each internal type with a corresponding versioned type. This, by itself, is not going to have any impact on users.
+### Introduce the "kubeadm alpha config view upgradeable" command
 
-### (OPTIONAL) “kubeadm upgrade plan” changes
-As supported component configs are expected to increase in number in the future and some of them might require manual user migration prior to upgrade, an ahead of time notice about what actions are required is a nice thing to have.  
-Hence, the “kubeadm upgrade plan” command is going to provide information on the component configs. In particular, if they are managed by kubeadm or are user supplied, and if they need manual upgrade or not.
+This new command would fetch the YAML documents of any component configs that are in use in the cluster and require manual migration to a newer version prior to cluster upgrade. Then it would print those YAML documents to the standard output or to a file, specified with the `--output` command line option.  
+If there are no configs that need manual upgrade, this command would print nothing.
```

Comment:
Is it expected that the output of this command can be passed to upgrade plan/apply?

Reply:
No. The output there is the set of old component configs that cannot be parsed by kubeadm.
The idea behind this command is for cluster admins to have a handy tool with which they can get the current contents of all component configs that need upgrading. That way they can take a note of what was modified by them in the first place and make sure the same changes are taken into account when creating the new component configs to be supplied to `kubeadm upgrade apply`.

On the other hand, the output of `kubeadm alpha config print component-defaults` is expected to be consumable directly by `kubeadm upgrade plan` and `kubeadm upgrade apply`.